src/components/mc/cuda/kernel/mc_cuda_reduce_multi.cu

```diff
@@ -0,0 +1,198 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2020-2021.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "../mc_cuda.h"
+#include "utils/ucc_math.h"
+#ifdef __cplusplus
+}
+#endif
+
+#include <assert.h>
+#include <stdio.h>
+
+#define  DO_OP_MAX(_v1, _v2) (_v1 > _v2 ? _v1 : _v2)
+#define  DO_OP_MIN(_v1, _v2) (_v1 < _v2 ? _v1 : _v2)
+#define  DO_OP_SUM(_v1, _v2) (_v1 + _v2)
+#define DO_OP_PROD(_v1, _v2) (_v1 * _v2)
+#define DO_OP_LAND(_v1, _v2) (_v1 && _v2)
+#define DO_OP_BAND(_v1, _v2) (_v1 & _v2)
+#define  DO_OP_LOR(_v1, _v2) (_v1 || _v2)
+#define  DO_OP_BOR(_v1, _v2) (_v1 | _v2)
+#define DO_OP_LXOR(_v1, _v2) ((!_v1) != (!_v2))
+#define DO_OP_BXOR(_v1, _v2) (_v1 ^ _v2)
+
+#define CUDA_REDUCE_WITH_OP(NAME, OP)                                          \
+template <typename T>                                                          \
+__global__ void UCC_REDUCE_CUDA_ ## NAME (const T *s1, const T *s2, T *d,      \
+                                          size_t size, size_t count,           \
+                                          size_t ld)                           \
+{                                                                              \
+        size_t start = blockIdx.x * blockDim.x + threadIdx.x;                  \
+        size_t step  = blockDim.x * gridDim.x;                                 \
+        for (size_t i = start; i < count; i+=step) {                           \
```

Comment:
General question (also applies to just reduce_cuda): each thread here performs reduction of values separated by "step" right? why is it better than having local "contiguous" piece of array to reduce?

Reply:
Correct, I think this is preferred way of doing reduction because of memory coalescing in warp